### PR TITLE
add support for network wasm filters

### PR DESCRIPTION
**Pull request by ramaraochavali**
Adds support for specifying plugin type

### CODE REVIEW

**Comment by ramaraochavali** [comment]
@kyessenov PTAL

File: type/v1beta1/selector.proto [hunk 1]
```diff
@@ -66,3 +69,14 @@ enum WorkloadMode {
   // source or destination of the network traffic.
   CLIENT_AND_SERVER = 3;
 }
+
+// ProtocolSelector is the criteria for specifying if a policy can be applied to
+// traffic of a specific protocol.
+enum ProtocolSelector {
+  // Applies Policy to HTTP traffic only.
+  HTTP = 0;
```
[the remark this answers is not in the document]
Comment (ramaraochavali) [reply type/v1beta1/selector.proto:77]: I was thinking of using HTTP to align with current default. But we can do that code.

[the remark this answers is not in the document]
Comment (kyessenov) [reply type/v1beta1/selector.proto]: I think it's fine to leave HTTP as the default. Saying UNDEFINED = HTTP indirectly adds to the confusion.

File: type/v1beta1/selector.proto [hunk 2]
```diff
@@ -41,6 +41,9 @@ message WorkloadSelector {
 message PortSelector {
   // Port number
   uint32 number = 1;
+
+  // Protocol used by the port
+  ProtocolSelector protocol = 2;
```
Comment (kyessenov) [inline type/v1beta1/selector.proto:46]: Does this force protocol handling on the port? I'm confused what it means to say port: 80, protocol: TCP on inbound, for example. Does that apply to HTTP traffic as well?

File: type/v1beta1/selector.proto [hunk 3]
```diff
@@ -66,3 +69,19 @@ enum WorkloadMode {
...
+// traffic of a specific protocol.
+enum ProtocolSelector {
+  // Default value, which will be interpreted by its own usage.
+  P_UNDEFINED = 0;
+
+  // Applies Policy to HTTP traffic only.
+  HTTP = 1;
+  
+  // Applies Policy to TCP traffic only.
+  TCP = 2;
```
Comment (kyessenov) [inline type/v1beta1/selector.proto:83]: I'd prefer us not relying on protocol sniffing at all. Instead, we should say that network filters apply on all ports (including HTTP), and then provide guidance how to detect HTTP traffic within Wasm.

Comment (ramaraochavali) [reply type/v1beta1/selector.proto:83]: Are you saying we ONLY apply network WASM (if a flag is chosen) on the selected port and if not apply HTTP default? But would n't that be too restrictive?

File: type/v1beta1/selector.proto (hunk 2 above)
Comment (kyessenov) [reply type/v1beta1/selector.proto:46]: HTTP is TCP based though? How do you apply network filters to HTTP chains then?

Comment (hzxuzhonghu) [inline type/v1beta1/selector.proto:46]: I am doubting how it works.  Say i specify a port 80, it is HTTP, but can I also specify ProtocolSelector = TCP?

Comment (costinm) [reply type/v1beta1/selector.proto:46]: We keep having those discussions. Next will be - what if HTTP is over QUIC and UDP ?

The problem is that we model the API on how envoy generates config, not what makes sense to user. 

IMO we should keep L4 policies separate from L7 policies - in particular given that L4 policies may someday be implemented by ztunnel. Running WASM in rust  is not hard - I know we're now saying "we'll never add features
to ztunnel" - but maybe some other implementation will handle L4 and support WASM, or maybe we change our mind. 

So maybe we not use 'protocol' - but keep separate L4 under a different name/resource (or some other mechanism), and always apply L4.

Comment (costinm) [reply type/v1beta1/selector.proto:46]: Agreed, it is confusing. That's why they defined the OSI layers...

File: type/v1beta1/selector.proto (hunk 3 above)
Comment (costinm) [reply type/v1beta1/selector.proto:83]: I don't know what would be restrictive about having deterministic and clear behavior. 

Completely agree - TCP policies don't require any sniffing ( obviously ), and for HTTP it is far better to be explicit, users typically know what ports are HTTP in particular if they have custom WASM code. 

100% of not adding any more APIs that depend or assume sniffing. In K8S we have HttpRoute and TcpRoute, there is no ambiguity.

File: type/v1beta1/selector.proto (hunk 2 above)
Comment (kyessenov) [reply type/v1beta1/selector.proto:46]: +1, that's what I was thinking, too. Network wasm API is essentially a copy of HTTP Wasm API that can bind to any port, just like HTTP Wasm API. We add a field at the very top of API to indicate it's "network" to save on creating another CR.

Network Wasm can bind to a port that has HTTP routes, that's sort of irrelevant from L4 perspective.

Comment (ramaraochavali) [reply type/v1beta1/selector.proto:46]: Are you suggesting add a flag at higher level some where in WasmPlugin message `enable_network_filter` (defaults to false existing behaviour) and if set, add it in the listener directly?
If yes, that sounds reasonable to me.

Comment (kyessenov) [reply type/v1beta1/selector.proto:46]: Yes, probably enum, with the default value HTTP_FILTER and a new value NETWORK_FILTER. In the future we can add LISTENER_FILTER, etc.

**Comment by ramaraochavali** [comment]
Updated to use Enum in the top level API that tells what type extension should be used

**Review by kyessenov — APPROVED** [review]

**Review by zirain — APPROVED** [review]

**Review by costinm — APPROVED** [review]

**Comment by ramaraochavali** [comment]
@howardjohn can you PTAL when you get chance?

**Review by hzxuzhonghu — APPROVED** [review]

File: extensions/v1alpha1/wasm.proto [hunk 4]
```diff
@@ -344,6 +344,19 @@ message WasmPlugin {
...
+
+  // Specifies the type of Wasm Extension to be used.
+  PluginType type = 14;
+}
+
+
+// PluginType indicates the type of Wasm Extension to be used.
+enum PluginType {
+  // Use HTTP Wasm Extension.
+  HTTP = 0;
```
Comment (howardjohn) [inline extensions/v1alpha1/wasm.proto:356]: Please make this `undefined = 0`; we may not need it now but it always comes back to byte us when we don't.

File: extensions/v1alpha1/wasm.proto [hunk 5]
```diff
@@ -344,6 +344,19 @@ message WasmPlugin {
...
   // the traffic passes the WasmPlugin.
   repeated TrafficSelector match = 12;
+
+  // Specifies the type of Wasm Extension to be used.
+  PluginType type = 14;
+}
+
+
+// PluginType indicates the type of Wasm Extension to be used.
+enum PluginType {
```
Comment (howardjohn) [inline extensions/v1alpha1/wasm.proto:354]: can we be a bit more descriptive one what these mean? Maybe link to the ABIs for HTTP and Network, whether network applies to HTTP as well or not, etc

**Comment by ramaraochavali** [comment]
@howardjohn added docs. PTAL

File: extensions/v1alpha1/wasm.proto [hunk 6]
```diff
@@ -344,6 +344,30 @@ message WasmPlugin {
...
+  // Specifies the type of Wasm Extension to be used.
+  PluginType type = 14;
+}
+
+
+// PluginType indicates the type of Wasm Extension to be used. 
+// There are two types of Extensions: HTTP and NETWORK.
+// HTTP Extension work at "Layer 7"(for example as a HTTP filters in Envoy). 
+// The detailed HTTP interface for can be found at [C++] (https://github.com/proxy-wasm/proxy-wasm-cpp-host/blob/b7e690703c7f26707438a2f1ebd7c197bc8f0296/include/proxy-wasm/context_interface.h#L199)
+// and [RUST] (https://github.com/proxy-wasm/proxy-wasm-rust-sdk/blob/6b47aec926bc29971c727471d6f4c972ec407c7f/src/traits.rs#L309).
```
Comment (howardjohn) [inline extensions/v1alpha1/wasm.proto:357]:
```suggestion
// and [Rus] (https://github.com/proxy-wasm/proxy-wasm-rust-sdk/blob/6b47aec926bc29971c727471d6f4c972ec407c7f/src/traits.rs#L309).
```

File: extensions/v1alpha1/wasm.proto [hunk 7]
```diff
@@ -344,6 +344,30 @@ message WasmPlugin {
...
+// NETWORK Extension work at "Layer 4"(for example as a Network filters in Envoy).
```
Comment (howardjohn) [inline extensions/v1alpha1/wasm.proto:358]:
```suggestion
// NETWORK Extension work at "Layer 4" (for example, as a network filter in Envoy).
```

File: extensions/v1alpha1/wasm.proto [hunk 8]
```diff
@@ -344,6 +344,30 @@ message WasmPlugin {
...
   repeated TrafficSelector match = 12;
+
+  // Specifies the type of Wasm Extension to be used.
+  PluginType type = 14;
+}
+
+
+// PluginType indicates the type of Wasm Extension to be used. 
+// There are two types of Extensions: HTTP and NETWORK.
+// HTTP Extension work at "Layer 7"(for example as a HTTP filters in Envoy).
```
Comment (howardjohn) [inline extensions/v1alpha1/wasm.proto:355]:
```suggestion
// HTTP Extension work at "Layer 7" (for example as an HTTP filter in Envoy). 
```

File: extensions/v1alpha1/wasm.proto [hunk 9]
```diff
@@ -344,6 +344,30 @@ message WasmPlugin {
...
   // the traffic passes the WasmPlugin.
   repeated TrafficSelector match = 12;
+
+  // Specifies the type of Wasm Extension to be used.
+  PluginType type = 14;
+}
+
+
+// PluginType indicates the type of Wasm Extension to be used. 
+// There are two types of Extensions: HTTP and NETWORK.
```
Comment (howardjohn) [inline extensions/v1alpha1/wasm.proto:354]:
```suggestion
// There are two types of Extensions: `HTTP` and `NETWORK`.
```

**Review by howardjohn — APPROVED** [review]
LGTM just some small comment suggestions

File: extensions/v1alpha1/wasm.proto (hunk 6 above)
Comment (zirain) [reply extensions/v1alpha1/wasm.proto:357]: rust

**Review by hanxiaop — APPROVED** [review]

**Pull request merged**